### PR TITLE
Fix #7294: Add limit on trees that can be inlined in a compilation unit

### LINKED ISSUE #7294: Infinite loop with OutOfMemory error while inlining givens
```Scala
trait Foo { def g(x: Int): Any }

inline given f[T <: Foo] as T = ??? match {
  case x: T => x.g(10)
}

@main def Test = f
```

Says:

<details><summary>Error</summary>

```scala
-- [E007] Type Mismatch Error: ../pg/i4.scala:4:18 -----------------------------
4 |  case x: T => x.g(10)
  |               ^^^^^^^
  |               Found:    Any
  |               Required: T
  |
  |               where:    T is a type in method f with bounds <: Foo
-- Error: ../pg/i4.scala:7:17 --------------------------------------------------
7 |@main def Test = f
  |                 ^
  |                 Maximal number of successive inlines (32) exceeded,
  |                 Maybe this is caused by a recursive inline method?
  |                 You can use -Xmax-inlines to change the limit.
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:7
-- [E005] Naming Error: ../pg/i4.scala:4:7 -------------------------------------
4 |  case x: T => x.g(10)
  |       ^^^^
  |       duplicate pattern variable: x
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:4
  | This location is in code that was inlined at i4.scala:7

longer explanation available when compiling with `-explain`
three errors found
[error] Nonzero exit code returned from runner: 1
[error] (dotty-compiler / Compile / runMain) Nonzero exit code returned from runner: 1
[error] Total time: 14 s, completed Sep 23, 2019 2:41:14 PM
```
</details>

Also, if you replace the definition of `trait Foo` with the following one (notice the parameter's type of `g` has changed):

```scala
trait Foo { def g(x: Any): Any }
```

Then the output is identical – up to the line "three errors found". The compiler hangs before this line and after waiting ~30 seconds, says:

<details><summary>OutOfMemoryError</summary>

```scala
java.lang.OutOfMemoryError: GC overhead limit exceeded while compiling ../pg/i5.scala
Exception in thread "main" java.lang.OutOfMemoryError: GC overhead limit exceeded
	at dotty.tools.dotc.typer.Inliner$$anon$5$$Lambda$810/1857126088.get$Lambda(Unknown Source)
	at java.lang.invoke.LambdaForm$DMH/314265080.invokeStatic_LL_L(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/1241276575.linkToTargetMethod(LambdaForm$MH)
	at dotty.tools.dotc.typer.Inliner$$anon$5.updateTermRefCounts$1(Inliner.scala:1180)
	at dotty.tools.dotc.typer.Inliner$$anon$5.traverse(Inliner.scala:1186)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1482)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1482)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1388)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1483)
	at dotty.tools.dotc.typer.Inliner$$anon$5.traverse(Inliner.scala:1191)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1482)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1482)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1380)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1483)
	at dotty.tools.dotc.typer.Inliner$$anon$5.traverse(Inliner.scala:1191)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1482)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1482)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1386)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1483)
	at dotty.tools.dotc.typer.Inliner$$anon$5.traverse(Inliner.scala:1191)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1482)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1482)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.apply$$anonfun$1(Trees.scala:1368)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator$$Lambda$525/795273218.apply(Unknown Source)
	at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:169)
	at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:165)
	at scala.collection.immutable.List.foldLeft(List.scala:79)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.apply(Trees.scala:1368)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1386)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1483)
	at dotty.tools.dotc.typer.Inliner$$anon$5.traverse(Inliner.scala:1191)
	at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1482)
```
</details>

Copy-pastable example for tests:

```scala
trait Foo { def g(x: Any): Any }

inline given f[T <: Foo]: T = ??? match {
  case x: T => x.g(10)
}

@main def Test = f
```

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Whit this change, executing `testCompilation tests/neg/i7294` takes 3s. Before it timed out or ran out of memory.

**Comment by nicolasstucki** [comment]
We do need both the inline depth check. The count does not grow fast enough to avoid a StackOverflow.

**Comment by nicolasstucki** [comment]
The current limit was chosen to allow all the current tests to compile out of the box.

File: compiler/src/dotty/tools/dotc/ast/tpd.scala [hunk 1]
```diff
@@ -1230,14 +1236,23 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
...
   def enclosingInlineds(implicit ctx: Context): List[Tree] =
     ctx.property(InlinedCalls).getOrElse(Nil)
 
+  /** Record inlined trees */
+  def addInlinedTrees(n: Int)(implicit ctx: Context): Unit =
+    ctx.property(InlinedTrees).foreach(_.count += n)
+
+  /** Check if the limit on the number of inlined trees has been reached */
+  def reachedInlinedTreesLimit(implicit ctx: Context): Boolean =
+    ctx.property(InlinedTrees).fold(0)(_.count) > ctx.settings.XmaxInlinedTrees.value
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/ast/tpd.scala:1254]:
```suggestion
ctx.property(InlinedTrees) match
  case Some(n) => n > ctx.settings.XmaxInlinedTrees.value
  case None => false
```
I think that's clearer than using fold.

**Pull request merged**